zerver/views/user_settings.py

```diff
@@ -42,6 +42,7 @@ def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpRes
         raise JsonableError(_("Email address changes are disabled in this organization."))
 
     do_change_user_delivery_email(user_profile, new_email)
+    do_set_confirmation_status(confirmation_key, Confirmation.EMAIL_CHANGE)
```

Comment:
No need of this. Instead do this in `get_object_from_key`. Also do the same for all other places where `get_object_from_key` is used.